### PR TITLE
DAI: Add timestamping, ASRC: Add control loop for DAI skew

### PR DESCRIPTION
**Pull request by singalsu**
This pull request for review purpose consists of commits. It completes the ASRC feature to usable state with non-synchronous DAI connection once merged.

- DAI: Add timestamping to DAI and HDA, SSP, DMIC drivers
- ASRC: Add DAI skew tracking and control loop
- ASRC: Topology updates

### CODE REVIEW

File: src/drivers/intel/cavs/ssp.c [hunk 1]
```diff
@@ -962,6 +963,177 @@ static int ssp_get_fifo(struct dai *dai, int direction, int stream_id)
...
+
+		/* Clear NTK to enable successive timestamps */
+		io_reg_write(tsctrl, TS_LOCAL_TSCTRL_NTK_BIT);
+	}
+
+	spin_unlock(dai->lock);
+
+	if (ntk) {
+#if 0
+		offs_clk = offs & MASK(15, 0); /* FRM remains zero SSP */
```
Comment (paulstelian97) [inline src/drivers/intel/cavs/ssp.c]: I presume this is a draft for now? I don't think we should have commented out code like this except temporarily.

Comment (singalsu) [reply src/drivers/intel/cavs/ssp.c]: Yes, that's correct. Use of these features is still open so I didn't remove the code. The DAI / FW arch for this change is the beef and why I want to get the feedback.

File: src/drivers/intel/cavs/ssp.c [hunk 2]
```diff
@@ -962,6 +963,174 @@ static int ssp_get_fifo(struct dai *dai, int direction, int stream_id)
...
+		return -EINVAL;
+	}
+
+	if (cfg->index > DAI_NUM_SSP_BASE + DAI_NUM_SSP_EXT - 1) {
+		trace_ssp_error("ssp_ts_config(): ");
+		return -EINVAL;
+	}
+
+	/* TODO, use some existing platform macro */
+	cfg->walclk_rate = 19200000;
```
Comment (singalsu) [inline src/drivers/intel/cavs/ssp.c]: I'd prefer to retrieve this from platform specific macros instead of this magic value but I'm not sure what to use.

File: src/drivers/intel/cavs/ssp.c (hunk 1 above)
Comment (paulstelian97) [reply src/drivers/intel/cavs/ssp.c]: What should happen with DAIs which simply don't have a hardware timestamp? (I'm not aware of any for the ESAI for example -- unless I missed something in the manual this one simply doesn't have a timestamp at all)

Comment (singalsu) [reply src/drivers/intel/cavs/ssp.c]: Good question, probably the xxx_ts_config() in the driver should return -EINVAL. Also I haven't tested what if ts_ops is NULL. I will try such scenarios with BYT platform.

File: src/drivers/intel/cavs/ssp.c (hunk 2 above)
Comment (paulstelian97) [reply src/drivers/intel/cavs/ssp.c]: Is this supposed to be the clock rate for whatever the SSP uses to generate its time stamps?

Comment (singalsu) [reply src/drivers/intel/cavs/ssp.c]: Also, the HW timer retrieve via DAI plus generic sample count can be preserved in ASRC. Or if no desire to wrap a HW timer under DAI there could be other service for ASRC to get the time reference.

File: src/audio/dai.c [hunk 3]
```diff
@@ -53,7 +53,7 @@ struct dai_data {
 	uint32_t stream_id;
 	struct dma_sg_config config;
 	struct comp_buffer *dma_buffer;
-
+	struct sof_timestamp_cfg ts_config;
```
Comment (lgirdwood) [inline src/audio/dai.c]: nitpick. no need for the sof_ prefix since this is only used internally and not in ABI.

**Comment by singalsu** [comment]
I will update the functionality in this PR to retrieve the sample count from HW.  PR #2110 enables the DMA LLP and now there's a steady increment of 48 seen every 1 ms for SSP master interface at 48 kHz rate as expected.

File: src/audio/dai.c (hunk 3 above)
Comment (singalsu) [reply src/audio/dai.c]: I forgot to do this, will do for next version!

File: src/drivers/intel/cavs/hda.c [hunk 4]
```diff
@@ -35,6 +43,106 @@ static int hda_get_fifo(struct dai *dai, int direction, int stream_id)
...
+	/* Set HDA timestamp registers */
+	uint32_t cdmas;
+	uint32_t addr = hda_ts_local_tsctrl_addr();
+
+	/* Set DMIC timestamp registers */
+	spin_lock(dai->lock);
+
+	/* Set CDMAS(4:0) to match DMA engine index and direction */
+	cdmas = TS_LOCAL_TSCTRL_CDMAS(cfg->dma_id |
+		(cfg->direction == SOF_IPC_STREAM_PLAYBACK ? BIT(4) : 0));
```
Comment (singalsu) [inline src/drivers/intel/cavs/hda.c]: @tlauda I'd need your help here. The sample count register remains zero for HDA in this version. I wonder if it could be the LLP not enabled for HDA DMA? Or just wrong DMA id here. It is also something to discuss if I should drop the HDA and DMIC part since for now at least they are not slave interfaces and should not drift vs. FW schedule timer.

The SSP and DMIC timestamps retrieve works perfectly, exactly as with settings you recommended for CDMAS.

**Comment by lgirdwood** [comment]
@singalsu any ETA for updates ?

**Comment by singalsu** [comment]
@lgirdwood I'm targeting to complete this work during this week. The opens are:

- Implement clean fail with non-cAVS / non-Intel platforms
- Find out the correct CDMAS programming for HD-audio or for-now remove HD-audio support

File: src/drivers/intel/cavs/dmic.c [hunk 5]
```diff
@@ -1583,6 +1584,97 @@ static int dmic_get_fifo(struct dai *dai, int direction, int stream_id)
...
+{
+	/* Read DMIC timestamp registers */
+	uint32_t ntk;
+	uint32_t addr;
+	uint32_t tsctrl = dmic_ts_local_tsctrl_addr();
+
+	/* Read SSP timestamp registers */
+	spin_lock(dai->lock);
+	ntk = io_reg_read(tsctrl) & TS_LOCAL_TSCTRL_NTK_BIT;
+	if (ntk) {
```
Comment (dbaluta) [inline src/drivers/intel/cavs/dmic.c]: better to fail early here and check for if (!ntk). This will also help you reducing one level of indentation.

File: src/drivers/intel/cavs/hda.c [hunk 6]
```diff
@@ -49,4 +157,9 @@ const struct dai_driver hda_driver = {
 		.probe			= hda_dummy,
 		.remove			= hda_dummy,
 	},
+	.ts_ops = {
+		.ts_config		= hda_ts_config,
+		.ts_start		= hda_ts_start,
+		.ts_get			= hda_ts_get,
```
Comment (dbaluta) [inline src/drivers/intel/cavs/hda.c]: Is there also a ts_stop?

File: src/audio/asrc/asrc.c [hunk 7]
```diff
@@ -487,26 +598,110 @@ static int asrc_prepare(struct comp_dev *dev)
...
+	int32_t f_ck_fs;
+	int ts_ret = -1;
+
+	if (cd->track_drift) {
+		ts_ret = asrc_dai_get_timestamp(cd, &tsd);
+		asrc_dai_start_timestamp(cd);
+	}
+
+	if (!ts_ret) {
+		ts = (int32_t)(tsd.walclk); /* Let it wrap, diff unwraps */
```
Comment (dbaluta) [inline src/audio/asrc/asrc.c]: You can save one indentation level if you do the if (ts_ret) return 0; first

File: src/drivers/intel/cavs/hda.c (hunk 6 above)
Comment (singalsu) [reply src/drivers/intel/cavs/hda.c]: The cAVS feature does not need a stop but it could be useful generally if there would be more to set up. The DMA LLPs are enabled now by default but it they weren't then start/stop would need to control them.

I feel it could be useful to add stop as preparation for future or for other platforms. What do you think @lgirdwood @dbaluta ?

File: src/audio/asrc/asrc.c (hunk 7 above)
Comment (singalsu) [reply src/audio/asrc/asrc.c]: Yep

Comment (singalsu) [reply src/audio/asrc/asrc.c]: I noticed the possible error from asrc_dai_get_timestamp() is not propagated. I'll fix it too.

File: src/audio/dai.c (hunk 3 above)
Comment (singalsu) [reply src/audio/dai.c]: Also "struct sof_timestamp_data" will be renamed, it's similar.

File: src/drivers/intel/cavs/dmic.c (hunk 5 above)
Comment (singalsu) [reply src/drivers/intel/cavs/dmic.c]: I did it this way to have a single spin_lock(); do_things(); spin_unlock(); flow. But OTOH I'm not sure using spin lock is needed here. I'm not sure that another DAI instance would mess things up if it could get in between these I/O operations. I'm a signal processing guy and not in my comfort zone with drivers. So I'd not change this but will do if the experts here confirm it's the right thing to do. OK?

File: src/audio/dai.c (hunk 3 above)
Comment (singalsu) [reply src/audio/dai.c]: Done!

**Comment by singalsu** [comment]
> @singalsu you probably also want to comment the ops and also state which are mandatory or optional.

Not yet done, the just added stop is optional for cAVS.

File: src/drivers/intel/cavs/hda.c (hunk 4 above)
Comment (singalsu) [reply src/drivers/intel/cavs/hda.c]: This is still open.

File: src/drivers/intel/cavs/hda.c (hunk 6 above)
Comment (singalsu) [reply src/drivers/intel/cavs/hda.c]: Done. I've added stop into the latest version. As suggested by Liam I should add note the it's optional.

File: src/audio/asrc/asrc.c (hunk 7 above)
Comment (singalsu) [reply src/audio/asrc/asrc.c]: Done.

This created a bit more change into ASRC like need like start of timestamp at prepare() and increase TS_STABLE_DIFF_COUNT by one for expected sane sample and clock deltas. It prevents the low-pass filter from being fed severely off values and cause xruns problem.

**Comment by singalsu** [comment]
I'll work next with what happens in non-cAVS platforms. I'll keep this work as [RFC] until that's checked.

**Comment by singalsu** [comment]
This version is checked with BYT SSP that doesn't support timestamps and updated to fail cleanly (no more DSP panic...) when playback is attempted. I've cleaned up also some extraneous traces from previous version.

**Comment by lgirdwood** [comment]
@singalsu what does fail cleanly mean here. Do we continue , XROn or STOP the pipeline ?

**Comment by singalsu** [comment]
> @singalsu what does fail cleanly mean here. Do we continue , XRUN or STOP the pipeline ?

I meant that the DSP doesn't crash and panic on platforms where e.g. the I2S driver has not set up timestamping function calls like BYT for SSP. The timestamp HW setup is done in ASRC case in prepare(). The prepare will return -ENXIO ( No such device or address) if that happens and e.g. aplay  report I/O error.  But other sound devices will continue to work.

The fail on BYT could be made to happen in topology instantation in ASRC new() with a bit more code if that's preferred.

ASRC can be run on BYT with the asynchronous mode token set to false (XRUNS will happen due to slow generic C version). This error happens if a topology requests ASRC  to track the pipeline DAI (asyncronous mode set to true) on platforms without timestamp HW capability or missing driver like in cAVS case for ALH and SDW. This PR supports SSP and DMIC and with a bug present HDA).

**Comment by lgirdwood** [comment]
@singalsu ok, so do you mean ASRC will work on BYT in SSP slave mode (i.e. codec/BT/modem is clocking the data) and ASRC can align audio between SSP (codec) time and DSP (host) time ?

**Comment by singalsu** [comment]
@lgirdwood Unfortunately not, ASRC could work in BYT only as alternate SRC implementation. There has been talk with @plbossart to add skew control from user space (e.g. match codec rate with USB audio class with user space based drift estimation). That would work with any platform.

The problem is with ALSA controls. You had idea to use mixer control with slider for 1s, 1/10s, 1/100s, 1/1000s. I haven't tried it yet but it should be possible.

File: src/drivers/intel/cavs/dmic.c [hunk 8]
```diff
@@ -1584,6 +1585,102 @@ static int dmic_get_fifo(struct dai *dai, int direction, int stream_id)
...
+	/* Request on demand timestamp */
+	io_reg_write(addr, TS_LOCAL_TSCTRL_ODTS_BIT | cdmas);
+
+	spin_unlock(dai->lock);
+	return 0;
+}
+
+static int dmic_ts_stop(struct dai *dai, struct timestamp_cfg *cfg)
+{
+	return 0;
```
Comment (lgirdwood) [inline src/drivers/intel/cavs/dmic.c]: Can we disable the TS, I'm suspicious leaving this ON may cause issues for D3.

File: src/drivers/intel/cavs/hda.c [hunk 9]
```diff
@@ -35,6 +43,110 @@ static int hda_get_fifo(struct dai *dai, int direction, int stream_id)
...
+	/* Request on demand timestamp */
+	io_reg_write(addr, TS_LOCAL_TSCTRL_ODTS_BIT | cdmas);
+
+	spin_unlock(dai->lock);
+	return 0;
+}
+
+static int hda_ts_stop(struct dai *dai, struct timestamp_cfg *cfg)
+{
+	return 0;
```
Comment (lgirdwood) [inline src/drivers/intel/cavs/hda.c]: ditto, lets disable this.

File: src/drivers/intel/cavs/dmic.c (hunk 8 above)
Comment (singalsu) [reply src/drivers/intel/cavs/dmic.c]: The ODTS bit is RWS type so it is cleared every time NTK bit is set. But it does not hurt to clear CDMA that is preserved as normal RW type bitfield. CDMAS links TS block to LLP count of the DMA controller/channel so it could cause something.

I'll add to new version an `io_reg_write(dmic_ts_local_tsctrl_addr(), 0);`

File: src/drivers/intel/cavs/hda.c (hunk 9 above)
Comment (singalsu) [reply src/drivers/intel/cavs/hda.c]: yep.

File: src/drivers/intel/cavs/dmic.c (hunk 8 above)
Comment (singalsu) [reply src/drivers/intel/cavs/dmic.c]: Done. I added also check into asrc_reset() check of cd->track_drift for executing asrc_dai_stop_timestamp();

File: src/drivers/intel/cavs/hda.c (hunk 9 above)
Comment (singalsu) [reply src/drivers/intel/cavs/hda.c]: Done!

File: src/drivers/intel/cavs/dmic.c (hunk 5 above)
Comment (lgirdwood) [reply src/drivers/intel/cavs/dmic.c]: @singalsu @dbaluta means something like this
```
spin_lock()
ntk = io_read(X);
if (!ntk)
    goto out;

/* set up clock since ntk is valid */

out:
spin_unlock();
```

Comment (singalsu) [reply src/drivers/intel/cavs/dmic.c]: OK I got it now. I guess I introduced "addr" to avoid long lines and now with less indent it's no more needed...

This change is in the latest pushed version.

**Comment by lgirdwood** [comment]
@singalsu can you monitor CI and rerun this PR when jenkins comes back on to get ICL results.

**Comment by singalsu** [comment]
SOFCI_TEST

**Comment by singalsu** [comment]
SOFCI TEST

File: src/drivers/intel/cavs/hda.c (hunk 4 above)
Comment (lgirdwood) [reply src/drivers/intel/cavs/hda.c]: @tlauda ping

File: src/audio/asrc/asrc.c [hunk 10]
```diff
@@ -354,6 +378,105 @@ static int asrc_params(struct comp_dev *dev,
 	return 0;
 }
 
+static int asrc_dai_find(struct comp_dev *dev,
```
Comment (lgirdwood) [inline src/audio/asrc/asrc.c]: This should be a pipeline API call. @bkokoszx I think this also aligns with your verify PR as you have similar logic for finding DAIs ? We should reuse the same function.

**Comment by singalsu** [comment]
I fixed the merge conflict and changed spin_lock/unlock() calls.

However there's a problem that if I play a list of files without pause between (manual aplay) the second and successive attempt fails to "pipeline_copy() error: ret = -61, start->comp.id = 6, dir = 1". I need to find out what causes it. Disabling the asynchronous mode avoids the error.

File: src/audio/asrc/asrc.c (hunk 10 above)
Comment (singalsu) [reply src/audio/asrc/asrc.c]: @bkokoszx Some kind of walk upstream/downstream until component type equals found function would be useful. In this case the walk starts from ASRC. Does your PR #2228 contain such already?

File: src/drivers/intel/cavs/hda.c (hunk 4 above)
Comment (tlauda) [reply src/drivers/intel/cavs/hda.c]: @singalsu @lgirdwood I'm pretty sure it should work. I can look at it as soon as I'm finished with my current effort.

File: src/audio/asrc/asrc.c (hunk 10 above)
Comment (bkokoszx) [reply src/audio/asrc/asrc.c]: @singalsu 
In PR #2228 (feature #1280) I've added function ```pipeline_comp_hw_params()```, which uses already implemented recursive function ```pipeline_for_each_comp()```. It goes from ```HOST``` to ```DAI``` end fetches DAI parameters set during ```dai_set_config()```. I will try to take a look how can we unify both patches.

Comment (singalsu) [reply src/audio/asrc/asrc.c]: Thanks @bkokoszx !

File: src/drivers/intel/cavs/hda.c (hunk 4 above)
Comment (singalsu) [reply src/drivers/intel/cavs/hda.c]: @tlauda I just re-run test with hda (topology sof-hda-asrc-2ch.tplg) and I get this error to trace `asrc.c:709 	asrc_control_loop(), DAI timestamp failed`. The time stamping programming sequence change for SSP DAI freeze as expected didn't improve things with HDA. I tried again to invert bit4 of CDMAS bit field here but no change.

When testing with playback the function hda_ts_start() seems to set every time CDMAS to value 0x17. The "1" is for playback direction and "7" for dma_id. The trace shows from elsewhere `hda-dmac: 7 channel 0 -> enable`. But I'm not sure this is the correct value to use here since it was more complicated for SSP.

Comment (tlauda) [reply src/drivers/intel/cavs/hda.c]: @singalsu I can successfully get samples when triggering timestamp. For HDA both directions should be 0 based in CDMAS, so something like this:
Link Input --> 0x0 for channel 0, 0x1 for channel 1, 0x2 for channel 2 etc.
Link Output --> 0x10 for channel 0, 0x11 for channel 1, 0x12 for channel 2 etc.

Comment (singalsu) [reply src/drivers/intel/cavs/hda.c]: Thanks!! Yes it works if I use the channel # here.

Comment (singalsu) [reply src/drivers/intel/cavs/hda.c]: I just pushed a new PR version with this fix. Now the DAI tracking works for HDA too. I noticed that when I attempt to play 44100 via ASRC to 48k HDA there's error:

```
   0      1          DMA         159132665.104167        12.916667 intel/cavs/hda-dma.c:653 	hda-dmac: 7 chan 0 - buffer not DMA aligned 0x2d0
    0      1         PIPE         159132689.114583        24.010416 src/audio/pipeline.c:503 	pipeline_prepare() error: ret = -22,dev->comp.id = 1
    0      1          IPC         159132708.281250        19.166666    src/ipc/handler.c:305 	ipc: pipe 1 comp 1 prepare failed -22
```
Maybe the possible odd number of frames produced causes it? I don't get this error with 48 kHz family rates.

**Comment by singalsu** [comment]
The DAI component discovery move to / use from pipeline is still missing from this PR version. Otherwise it looks pretty good. I will work meanwhile with another pull mode PR for ASRC. It has so far used only the push mode filter. I've totally forgotten to implement use of pull mode (for capture).

**Comment by singalsu** [comment]
Gah, the ASRC in git master has become broken, bisecting...

**Comment by singalsu** [comment]
I pushed a rebased version with merge conflict fixed. However the regression is still there. The problem is that first playback after boot on my UP2 fails into "aplay: main:788: audio open error: Device or resource busy". Successive playback attempts work OK.

The previous rebase 3 hours ago did not have this issue.

**Comment by lgirdwood** [comment]
@singalsu have you checked the trace ? There should be a reason why it fails ?

**Comment by singalsu** [comment]
> @singalsu have you checked the trace ? There should be a reason why it fails ?

@lgirdwood It was my own mistake! The UP2 offered an upgrade that I accepted and it started pulseaudio that is normally renamed to avoid it to execute. Now pulseaudio kept the audio device busy in my first play attempt.

**Comment by singalsu** [comment]
Now the PR works OK here. I'll still update the DAI find part with @bkokoszx .

**Comment by bkokoszx** [comment]
@singalsu 
Sorry, but I have still one problem with my patch (#2228). As soon as I finish I will take care of ```arc_find_dai()``` function.

**Comment by singalsu** [comment]
I just updated this PR due to need to fix merge conflict in dai.c (cache function removal). There's no other changes.

File: src/drivers/intel/cavs/hda.c (hunk 4 above)
Comment (singalsu) [reply src/drivers/intel/cavs/hda.c]: @tlauda @mmaka1 The same ASRC pipeline macro runs OK with 44100 input in APL and I2S codec. I don't think the ASRC should have responsibility to ensure output size to PCM buffer is N * 32 bytes. The error above looks like HDA codec driver issue to me.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
@singalsu @bkokoszx I'll merge this now, we can update the common API when @bkokoszx is ready.

**Comment by singalsu** [comment]
Woohoo, thanks to everyone for review work! This one was tricky and I would have not been able to do this without your help.
